### PR TITLE
keras support added

### PR DESCRIPTION
**Pull request by gauravgupta22**
@virajprabhu This does not uses https://github.com/MarcBS/keras. 
MarcBS/keras do not take into account the differences b/w caffe and tensorflow when applying pooling. Also it does not support 'same' padding.
Our implementation covers both these limitations
It work by converting Keras -> Json and Json -> Keras
You can always export any model to keras
But since the IDE is caffe centered, some keras models cannot be imported

### CODE REVIEW

**Comment by virajprabhu** [comment]
@gauravgupta22 great! Sounds reasonable. I'll review this soon.

File: README.md [hunk 1]
```diff
@@ -2,13 +2,14 @@
 
 This is a React+Django webapp with a simple drag and drop interface to build and configure deep neural networks with support for export of model configuration files to caffe and tensorflow. It also supports import from these frameworks to visualize different model architectures. Our motivation is to build an online IDE where researchers can share models and collaborate without having to deal with deep learning code.
 
-![alt text](https://github.com/Cloud-CV/IDE/blob/master/example/snapshot.png "IDE Snapshot")
+![alt text](https://github.com/gauravgupta22/IDE/blob/master/example/snapshot.png "IDE Snapshot")
```
Comment (deshraj) [inline README.md]: @gauravgupta22 I think the link should belong to the CloudCV repository only.

File: keras_app/urls.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+from django.conf.urls import url
+from views.import_keras import import_keras_json
```
Comment (deshraj) [inline keras_app/urls.py]: Can you replace it with `from .views.import_keras import import_keras_json`?

File: keras_app/urls.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+from views.export_keras import export_keras_json
```
Comment (deshraj) [inline keras_app/urls.py]: Can ou replace it with `from .views.export_keras import export_keras_json`

File: keras_app/views/__init__.py [hunk 4]
```diff
@@ -0,0 +1,2 @@
+import import_keras
+import export_keras
```
Comment (deshraj) [inline keras_app/views/__init__.py]: I don't think we need it. @gauravgupta22 please correct me if I am wrong.

File: keras_app/views/export_keras.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
+from django.shortcuts import render
+from django.views.decorators.csrf import csrf_exempt
+from django.http import JsonResponse
+from json2keras import json_to_keras
+import yaml
+from ide.utils.json_utils import preprocess_json
+from datetime import datetime
+import random, string
+import os
```
Comment (deshraj) [inline keras_app/views/export_keras.py]: Can we have imports sorted in alphabetical order?

File: keras_app/views/export_keras.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+BASE_DIR = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+
+def randomword(length):
+    return ''.join(random.choice(string.lowercase) for i in range(length))
+
+@csrf_exempt
```
Comment (deshraj) [inline keras_app/views/export_keras.py]: Should we really exempt? This doesn't look like a good idea.

File: keras_app/views/import_keras.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
+from django.shortcuts import render
+from django.views.decorators.csrf import csrf_exempt
+from django.http import JsonResponse
+from keras2json import keras_to_json
+import json
+
+@csrf_exempt
```
Comment (deshraj) [inline keras_app/views/import_keras.py]: Here also.

File: keras_app/views/json2keras.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
+import math
+import json
+from keras.layers.advanced_activations import *
+from keras.layers.convolutional import *
+from keras.layers.core import *
+from keras.layers.normalization import *
+from keras.layers import *
+from keras.models import Model
+from extra_layers import *
+from ide.utils.json_utils import *
```
Comment (deshraj) [inline keras_app/views/json2keras.py]: Can you please sort them lexicographical order?

File: tensorflow_app/views/export_graphdef.py [hunk 9]
```diff
@@ -30,7 +36,7 @@ def exportToTensorflow(request):
         os.system('python '+BASE_DIR+'/tensorflow_app/caffe-tensorflow/convert.py '+BASE_DIR+'/media/'+randomId+'.prototxt --code-output-path='+BASE_DIR+'/media/'+randomId+'.py')
```
Comment (deshraj) [inline tensorflow_app/views/export_graphdef.py]: I think we need to change this? `os.system()` is not the best way to do this. We can make it a python module and import the methods defined and then use them. Does that make sense?

**Comment by virajprabhu** [comment]
@gauravgupta22 You said "But since the IDE is caffe centered, some keras models cannot be imported" - can you list what can't be imported? We are trying to document these limitations.

**Comment by gauravgupta22** [comment]
There are two major issues when converting from keras => caffe :

- Differences in implementation of Pooling - In keras, the half-windows are discarded. Caffe will put additional output for half-windows. 
- Differences in Padding schemes - The 'same' padding in keras can sometimes result in different padding values for top-bottom (or left-right). caffe always pad evenly on both sides so the top-bottom (or left-right) padding values are always equal.

Depending on the parameters of the layer whenever we have a conflict on these issues we can't import.

Have a look at these links for more info :

- https://github.com/Yangqing/caffe2/blob/master/caffe2/proto/caffe2_legacy.proto
- http://joelouismarino.github.io/blog_posts/blog_googlenet_keras.html
- https://www.tensorflow.org/versions/r0.10/api_docs/python/nn.html

**Comment by deshraj** [comment]
@gauravgupta22 is this PR ready for merging?

**Comment by virajprabhu** [comment]
@deshraj I didn't get time to finish reviewing. Will finish and merge by end of the week.

File: ide/static/js/topBar.js [hunk 10]
```diff
@@ -18,6 +18,7 @@ class TopBar extends React.Component {
                   <ul className="dropdown-menu pull-right">
                     <li><a href="#" onClick={() => this.props.exportNet('caffe')}>caffe</a></li>
                     <li><a href="#" onClick={() => this.props.exportNet('tensorflow')}>tensorflow</a></li>
+                    <li><a href="#" onClick={() => this.props.exportNet('keras')}>keras</a></li>
```
Comment (virajprabhu) [inline ide/static/js/topBar.js]: Capitalize first letter of all frameworks.

File: ide/utils/jsonToPrototxt.py [hunk 11]
```diff
@@ -151,7 +154,8 @@ def changeTopBlobName(layerId, newName):
 
             if layerPhase is not None:
                 caffeLayer = get_iterable(L.Input(
-                    input_param={'shape':{'dim':map(int,layerParams['dim'].split(','))}},
+                    name = layerName,
+                    input_param={'shape':{'dim':layerParams['dim']}},
```
Comment (virajprabhu) [inline ide/utils/jsonToPrototxt.py]: Fix indentation

File: ide/utils/jsonToPrototxt.py [hunk 12]
```diff
@@ -166,28 +170,18 @@ def changeTopBlobName(layerId, newName):
             else:
                 for ns in (ns_train,ns_test):
                     caffeLayer = get_iterable(L.Input(
-                        input_param={'shape':{'dim':map(int,layerParams['dim'].split(','))}}))
+                        name = layerName,
+                        input_param={'shape':{'dim':layerParams['dim']}}))
```
Comment (virajprabhu) [inline ide/utils/jsonToPrototxt.py]: Fix indentation.

File: ide/utils/jsonToPrototxt.py [hunk 13]
```diff
@@ -295,6 +276,7 @@ def changeTopBlobName(layerId, newName):
                 caffeLayer = get_iterable(L.Accuracy(
                     *([ns[x] for x in blobNames[layerId]['bottom']]),
                     #*([ns[x] for x in blobNames[layerId]['bottom']] + [ns.label]),
```
Comment (virajprabhu) [inline ide/utils/jsonToPrototxt.py]: Remove unusable commented code everywhere.

**Comment by deshraj** [comment]
@gauravgupta22 any updates on this PR?

**Comment by virajprabhu** [comment]
Closing this due to no response. Changes pushed to [keras-support](https://github.com/Cloud-CV/IDE/tree/keras-support).
